### PR TITLE
Hypershift: Respect publishing strategy of OVN southbound database service

### PR DESCRIPTION
**Pull request by kyrtapz**

### CODE REVIEW

File: bindata/network/ovn-kubernetes/managed/008-route.yaml [hunk 1]
```diff
@@ -15,3 +15,6 @@ spec:
   to:
     kind: Service
     name: ovnkube-master
+{{ if .OVNSbDbRouteHost }}
+  host: {{.OVNSbDbRouteHost}}
```
Comment (zshi-redhat) [inline bindata/network/ovn-kubernetes/managed/008-route.yaml:19]: Did you see any issue by specifying route host? I remember I got permission issue when host is specified.

File: bindata/network/ovn-kubernetes/managed/008-route.yaml [hunk 2]
```diff
@@ -15,3 +15,6 @@ spec:
   to:
     kind: Service
     name: ovnkube-master
+{{ if .OVNSbDbRouteHost }}
```
Comment (squeed) [inline bindata/network/ovn-kubernetes/managed/008-route.yaml:18]: Presumably we should also avoid creating this Route at all if it's not needed?

Comment (kyrtapz) [reply bindata/network/ovn-kubernetes/managed/008-route.yaml:18]: Absolutely, missed that

File: pkg/network/ovn_kubernetes.go [hunk 3]
```diff
@@ -391,34 +405,76 @@ func bootstrapOVNHyperShiftConfig(hc HyperShiftConfig, kubeClient cnoclient.Clie
...
+	for _, svc := range hcp.Spec.Services {
+		// TODO: instead of the hardcoded string use ServiceType hyperv1.OVNSbDb once the API is updated
+		if svc.Service == "OVNSbDb" {
+			ovnHypershiftResult.ServicePublishingStrategy = &svc.ServicePublishingStrategy
 		}
-		if route.Spec.Host == "" && route.Status.Ingress[0].Host == "" {
-			return ovnHypershiftResult, nil
+	}
+	if ovnHypershiftResult.ServicePublishingStrategy == nil {
+		return nil, fmt.Errorf("service publishing strategy for OVN southband database does not exist in hyperv1.HostedControlPlane %s/%s", hc.Name, hc.Namespace)
```
Comment (squeed) [inline pkg/network/ovn_kubernetes.go:424]: Should we generate a reasonable default if this is missing? Presumably we can just default to Route.

**Comment by zshi-redhat** [comment]
It seems hypershift cno doesn't have clusterrole permission to set route host:
```
- lastTransitionTime: "2022-04-13T03:58:41Z"
  message: 'Error while updating operator configuration: could not apply (route.openshift.io/v1,
    Kind=Route) clusters-ovn1-hypershift/ovnkube-sbdb: failed to apply / update (route.openshift.io/v1,
    Kind=Route) clusters-ovn1-hypershift/ovnkube-sbdb: Route.route.openshift.io "ovnkube-sbdb"
    is invalid: spec.host: Forbidden: you do not have permission to set the host field
    of the route'
  reason: ApplyOperatorConfig
  status: "True"
```

**Comment by kyrtapz** [comment]
> It seems hypershift cno doesn't have clusterrole permission to set route host:
> 
> ```
> - lastTransitionTime: "2022-04-13T03:58:41Z"
>   message: 'Error while updating operator configuration: could not apply (route.openshift.io/v1,
>     Kind=Route) clusters-ovn1-hypershift/ovnkube-sbdb: failed to apply / update (route.openshift.io/v1,
>     Kind=Route) clusters-ovn1-hypershift/ovnkube-sbdb: Route.route.openshift.io "ovnkube-sbdb"
>     is invalid: spec.host: Forbidden: you do not have permission to set the host field
>     of the route'
>   reason: ApplyOperatorConfig
>   status: "True"
> ```

Addressed here: https://github.com/openshift/hypershift/pull/1253/files#diff-06207575f3cfe4069b485056885b8807a7ad0f29e07948aa573434531aa9665cR132

File: pkg/network/ovn_kubernetes.go [hunk 4]
```diff
@@ -415,35 +428,77 @@ func bootstrapOVNHyperShiftConfig(hc *HyperShiftConfig, kubeClient cnoclient.Cli
...
+				Version:  "v1",
+				Resource: "routes",
+			}
+			clusterClient := kubeClient.ClientFor(client.ManagementClusterName)
+			routeObj, err := clusterClient.Dynamic().Resource(gvr).Namespace(hc.Namespace).Get(context.TODO(), "ovnkube-sbdb", metav1.GetOptions{})
+			if err != nil {
+				if apierrors.IsNotFound(err) {
+					klog.Infof("Did not find ovnkube-sbdb route")
+				} else {
+					return nil, fmt.Errorf("could not get ovnkube-sbdb route: %v", err)
```
Comment (zshi-redhat) [inline pkg/network/ovn_kubernetes.go:458]: nit: return `ovnHypershiftResult` and err?

File: pkg/network/ovn_kubernetes.go [hunk 5]
```diff
@@ -415,35 +428,77 @@ func bootstrapOVNHyperShiftConfig(hc *HyperShiftConfig, kubeClient cnoclient.Cli
...
+			}
 		}
-		if route.Status.Ingress[0].Host != "" {
-			ovnHypershiftResult.Route = route.Status.Ingress[0].Host
-		} else if route.Spec.Host != "" {
-			ovnHypershiftResult.Route = route.Spec.Host
+	case hyperv1.NodePort:
+		{
+			svc := &corev1.Service{}
+			err = kubeClient.Default().CRClient().Get(context.TODO(), types.NamespacedName{Namespace: hc.Namespace, Name: "ovnkube-master"}, svc)
```
Comment (zshi-redhat) [inline pkg/network/ovn_kubernetes.go:479]: should this be ovnkube-master-external service?

Comment (kyrtapz) [reply pkg/network/ovn_kubernetes.go]: yes! done

File: pkg/network/ovn_kubernetes.go (hunk 4 above)
Comment (kyrtapz) [reply pkg/network/ovn_kubernetes.go:458]: Went the other way and stopped returning `ovnHypershiftResult` on error as it is not used anyway.

**Pull request merged**